php/libraries/FileUploadValidator.class.inc

```diff
@@ -0,0 +1,241 @@
+ *  @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ *  @link     https://www.github.com/aces/Loris
+ */
+
+/**
+ * Class to validate file uploads to LORIS.
+ *
+ *  @category Main
+ *  @package  Behavioural
+ *  @author   John Saigle <john.saigle@mcin.ca>
+ *  @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ *  @link     https://www.github.com/aces/Loris
+ */
+class FileUploadValidator
+{
+
+    public $uploadDirectory;
+    public $permittedMIMETypes;
+    private $_responseCode = 500;
+
+    /**
+     * Create new instance of a File Upload Validator.
+     *
+     * @param string $uploadDirectory    The target parent directory of files
+     *                              uploaded using this instance.
+     * @param array  $permittedMIMETypes A whitelist of all MIME types that can
+     * be accepted by this instance. E.g. array('image/png', 'text/plain')
+     *
+     * @return void
+     */
+    public function __construct(
+        string $uploadDirectory,
+        array $permittedMIMETypes = array()
+    ) {
+        $this->uploadDirectory    = $uploadDirectory;
+        $this->permittedMIMETypes = $permittedMIMETypes;
+        $this->validate();
+    }
+
+    /**
+     * Validates the constructor by ensuring the upload path exists and is
+     * writable.
+     *
+     * @throws \LorisException
+     *
+     * @return void
+     */
+    protected function validate(): void
+    {
+        if (empty($this->uploadDirectory)) {
+            throw new \InvalidArgumentException(
+                'Upload directory cannot be empty'
+            );
+        }
+        if (! is_dir($this->uploadDirectory)) {
+            $this->_responseCode = 500;
+            throw new \LorisException(
+                'Upload directory is not a directory'
+            );
+        }
+        if (! is_writable($this->uploadDirectory)) {
+            // Set __responseCode anyway in case some calling code decides to
+            // catch this exception.
+            $this->_responseCode = 500;
+            throw new \LorisException(
+                'Upload directory is not writable'
+            );
+        }
+        return;
+    }
+
+    /**
+     * Given an array of files, validate them and upload them to the server.
+     *
+     * The permission parameter is a UNIX file permissions octet.
+     * All files uploaded will use these permissions. If uploaded files are
+     * required to have differing permissions, this must be done
+     * manually in the calling code after this function has executed.
+     *
+     * If the $overwrite parameter is set to true, existing files will be
+     * overwritten.
+     *
+     * @param UploadedFile[] $uploadedFiles An array of files uploaded to the
+     *                              server. These files must implement the PSR-7
+     *                              UploadedFile interface.
+     * @param int            $permissions   A UNIX permissions octet (e.g. 0644)
```

Comment:
I think this and the below variables make more sense as parameters because you might want to upload multiple files having different permissions, I think? It makes a bit more sense to me to say "Object, create a file with these permissions" rather than use an instance variable to put the object into "0644" mode. 

What do you think?